File: src/main/java/gargoyle/ct/ClasspathResource.java
package gargoyle.ct;

import java.io.IOException;
import java.net.URL;

public class ClasspathResource extends VirtualResource {

    private ClasspathResource(Resource base, String location) {
        super(base, location);
    }

    public ClasspathResource(String location) {
        super(location);
    }

    @Override
    protected ClasspathResource createResource(Resource base, String location) {
        return base == null ? new ClasspathResource(location) : new ClasspathResource(base, location);
    }

    @Override
    public boolean exists() {
        try {
            return exists(toURL());
        } catch (IOException ex) {
            return false;
        }
    }

    @Override
    public URL toURL() throws IOException {
        return Thread.currentThread().getContextClassLoader().getResource(getLocation());
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/resource/internal/ClasspathResource.kt
package gargoyle.ct.util.resource.internal

import gargoyle.ct.util.resource.Resource
import java.net.URL

/**
 * represents a [Resource] withing application class path (ex. in jar file). writing is mainly unsupported
 */
open class ClasspathResource : VirtualResource<ClasspathResource> {
    val loader: ClassLoader

    constructor(location: String) : this(ClasspathResource::class.java.classLoader, location)
    constructor(loader: ClassLoader, location: String) : super(location) {
        this.loader = loader
    }

    internal constructor(loader: ClassLoader, base: ClasspathResource, location: String) : super(base, location) {
        this.loader = loader
    }

    override fun createResource(
        base: VirtualResource<ClasspathResource>?,
        location: String
    ): VirtualResource<ClasspathResource> =
        if (base == null) ClasspathResource(loader, location)
        else ClasspathResource(getLoader(base), base as ClasspathResource, location)

    override fun toURL(): URL = loader.getResource(location)!!

    fun getLoader(resource: Resource?): ClassLoader = if (resource is ClasspathResource) resource.loader else loader

    override fun exists(): Boolean =
        try {
            exists(loader.getResource(location))
        } catch (ex: Exception) {
            false
        }
}
